### PR TITLE
O3-898: Improve the search E2E test

### PR DESCRIPTION
**Pull request by kumuditha-udayanga**
## Purpose
The purpose of this PR is to fix [O3-898](https://issues.openmrs.org/browse/O3-898)

## Approach
- Create a new patient with a random name at the beginning of the test.
- Search for the newly created patient.
- Delete the patient at the end of the test.

### CODE REVIEW

File: cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js [hunk 1]
```diff
@@ -1,6 +1,22 @@
-import {Given} from 'cypress-cucumber-preprocessor/steps';
+import {Given, When, Before, After} from 'cypress-cucumber-preprocessor/steps';
 
-Given('the user login to the Registration Desk', () => {    
+let identifier = null;
+let patient = null;
+let patientName = `Patient${Math.random().toString(36).substring(2,7)}`;
+
+
```
Comment (kdaud) [inline cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:7]: isn't one empty line enough to separate declared variable from methods?
```suggestion
```

Comment (kumuditha-udayanga) [reply cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:7]: Yes @kdaud, I'll remove it

**Comment by kdaud** [comment]
have you taken a look at empty lines [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/115a62fab0a4248a11eda5010a885317d9884f65/cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js#L6-L7)?

File: cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js (hunk 1 above)
Comment (kdaud) [reply cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:7]: have you taken a look at the empty lines [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/115a62fab0a4248a11eda5010a885317d9884f65/cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js#L6-L7)?

Comment (kumuditha-udayanga) [reply cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js]: I removed the extra line, as suggested

**Comment by kdaud** [comment]
@jayasanka-sack what do you say about this PR?

File: cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js [hunk 2]
```diff
@@ -9,14 +24,19 @@ Given('the user login to the Registration Desk', () => {
     cy.visit('home');
 })
 
-When('the user search for {string}', patientName => {
+When('the user search for {string}', () => {
```
Comment (jayasanka-sack) [inline cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:27]: We don't need this parameter.
```suggestion
When('the user searches for a patient by name', () => {
```

File: src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature [hunk 3]
```diff
@@ -3,10 +3,7 @@ Feature: Patient Search
...
-  Scenario Outline: Search for a patient
-    When the user search for "<patientName>"
-    Then the result should be "<result>"
-    Examples:
-      | patientName | result           |
-      | Kevin Jones | Found 1 patient  |
-      | 100MQ       | No results found |
+  @patient-search
+  Scenario: Search for a patient
+    When the user search for "Patient"
```
Comment (jayasanka-sack) [inline src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature:8]:
```suggestion
    When the user searches for a patient by name
```

File: cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js [hunk 4]
```diff
@@ -9,14 +24,19 @@ Given('the user login to the Registration Desk', () => {
...
     // Adding an artificial wait because typed texts takes some time to get updated
     cy.wait(1000);
     cy.getByPlaceholder('Search for a patient by name or identifier number').type('{enter}');
 })
 
-Then('the result should be {string}', result => {
-    cy.contains(result);
+Then(`the result should be {string}`, () => {
+    cy.contains('Found 1 patient');
+});
```
Comment (jayasanka-sack) [inline cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:38]:
```suggestion
Then('the result should be {string}', result => {
    cy.contains(result);
});
```

**Review by jayasanka-sack — CHANGES_REQUESTED** [review]
Thanks @kumuditha-udayanga for the PR. I added some comments.

**Comment by kumuditha-udayanga** [comment]
@jayasanka-sack fixed the requested changes.

File: cypress/support/commands.js [hunk 5]
```diff
@@ -295,13 +295,13 @@ Cypress.Commands.add('generateLabResults', (uuid) => {
     });
 })
 
-Cypress.Commands.add('createPatient', (identifier = null) => {
+Cypress.Commands.add('createPatient', (identifier = null, name) => {
```
Comment (jayasanka-sack) [inline cypress/support/commands.js:298]: Let's make it more readable. (remove 'john' from line 304)
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters

```suggestion
Cypress.Commands.add('createPatient', (identifier = null, name = 'John') => {
```

File: cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js [hunk 6]
```diff
@@ -9,14 +24,19 @@ Given('the user login to the Registration Desk', () => {
...
+    cy.log(`Searching for ${patientName}`);
     cy.get('button[name=SearchPatientIcon]').click();
     cy.getByPlaceholder('Search for a patient by name or identifier number').type(patientName);
     // Adding an artificial wait because typed texts takes some time to get updated
     cy.wait(1000);
     cy.getByPlaceholder('Search for a patient by name or identifier number').type('{enter}');
 })
 
-Then('the result should be {string}', result => {
+Then(`the result should be {string}`, result => {
```
Comment (jayasanka-sack) [inline cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:36]:
```suggestion
Then('the result should be {string}', result => {
```

**Review by jayasanka-sack — CHANGES_REQUESTED** [review]

File: src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature [hunk 7]
```diff
@@ -3,10 +3,7 @@ Feature: Patient Search
   Background:
     Given the user login to the Registration Desk
 
-  Scenario Outline: Search for a patient
-    When the user search for "<patientName>"
-    Then the result should be "<result>"
-    Examples:
```
Comment (sherrif10) [inline src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature:9]: isnt the user supposed to enter patient to be searched here

File: cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js [hunk 8]
```diff
@@ -9,14 +24,19 @@ Given('the user login to the Registration Desk', () => {
     cy.visit('home');
 })
 
-When('the user search for {string}', patientName => {
+When('the user searches for a patient by name', () => {
+    cy.log(`Searching for ${patientName}`);
```
Comment (sherrif10) [inline cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:28]:
```suggestion
    cy.log('Searching for ${patientName}');
```

File: src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature (hunk 7 above)
Comment (kumuditha-udayanga) [reply src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature:9]: It was, from this PR the patient name is randomly generated, and we use that patient name throughout the rest of the test and delete the patient at the end of the test.

**Comment by sherrif10** [comment]
Hello @jayasanka-sack  help  us to approve  this pr. Thanks.

File: cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js [hunk 9]
```diff
@@ -1,6 +1,21 @@
...
+let patient = null;
+let patientName = `Patient${Math.random().toString(36).substring(2,7)}`;
+
+Before({tags: '@patient-search'}, () => {
+    cy.generateIdentifier().then((generatedIdentifier) => {
+        identifier = generatedIdentifier;
+        cy.createPatient(identifier, patientName).then((generatedPatient) => {
+            patient = generatedPatient;
+            cy.startFacilityVisit(patient.uuid);
+            cy.generateLabResults(patient.uuid);
```
Comment (jayasanka-sack) [inline cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js:13]: Could you please remove these two?

Comment (kumuditha-udayanga) [reply cypress/integration/cucumber/step_definitions/refapp-3.x/06-patient-search/patient-search.js]: Done

**Comment by sherrif10** [comment]
Seems the ticket title isn't enough to describe what's being achieved in the commit

File: src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature (hunk 7 above)
Comment (sherrif10) [reply src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature:9]: Sounds better.thanks

**Pull request closed without merging**